src/applications/vaos/hooks/useFormState.jsx

```diff
@@ -0,0 +1,90 @@
+/**
+ * @module hooks
+ */
+import { useState, useEffect, useRef } from 'react';
+import { getDefaultFormState } from '@department-of-veterans-affairs/react-jsonschema-form/lib/utils';
+
+import {
+  updateSchemaAndData,
+  updateItemsSchema,
+} from 'platform/forms-system/src/js/state/helpers';
+
+function setupFormData(data, schema, uiSchema) {
+  const schemaWithItemsCorrected = updateItemsSchema(schema);
+  return updateSchemaAndData(
+    schemaWithItemsCorrected,
+    uiSchema,
+    getDefaultFormState(schemaWithItemsCorrected, data, {}),
+  );
+}
+
+/**
+ * Hook for managing the state for a SchemaForm page in local component state
+ * instead of redux
+ *
+ * @export
+ * @name useFormState
+ * @function
+ * @param {Object} params Hook parameters
+ * @param {Object|Function} params.initialSchema The initial schema for the form. Can also be a function, which will
+ *   only be executed on startup or when dependencies changes
+ * @param {Object} params.uiSchema The uiSchema for the form
+ * @param {Object} params.initialData The initial data to use for the form state. This generally just used for the data
+ *   initially loaded into state, and any later changes are ignored. However if a dependency change causes the schema
+ *   to be re-generated, the most recent initialData passed in will be used, unless setData has been called
+ * @param {Array} [params.dependencies=[]] Array of useEffect dependencies that will force the form state to be reset using the current
+ *   values for initialSchema and uiSchema.
+ * @returns {{ schema: Object, uiSchema: Object, data: Object }} The form state values, to be used with SchemaForm
+ */
+export default function useFormState({
+  initialSchema,
+  uiSchema = null,
+  initialData = {},
+  dependencies = [],
+}) {
+  const initialLoadRef = useRef(true);
+  const dataUpdatedRef = useRef(false);
+  const [formState, setFormState] = useState(() => {
+    let schema = initialSchema;
+    if (typeof initialSchema === 'function') {
+      schema = initialSchema();
+    }
+
+    if (!schema) {
+      return { schema, data: initialData };
+    }
+
+    return setupFormData(initialData, schema, uiSchema);
+  });
+
+  useEffect(() => {
+    if (dependencies?.length && !initialLoadRef.current) {
+      let schema = initialSchema;
+      if (typeof initialSchema === 'function') {
+        schema = initialSchema();
+      }
+      setFormState(
+        setupFormData(
+          dataUpdatedRef.current ? formState.data : initialData,
```

Comment:
I was trying to handle cases where data might get updated through a Redux action after the first render. I could have a useEffect that looked for `initialData` changes, but that puts someone at risk of creating an infinite loop if they happen to pass something to `initialData` that has a different reference each time.